File: server/app/schemas/equipment.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List
from .manual import ManualContentResponse


class AttachmentResponse(BaseModel):
    id: int
    name: str
    type: str
    url: str
    thumbnail: Optional[str] = None
    metadata: Optional[dict] = Field(default=None, alias="file_metadata")
    upload_date: datetime
    document_category: str = "implementation"
    is_published: bool = True

    class Config:
        from_attributes = True
        populate_by_name = True


class EquipmentCreate(BaseModel):
    name: str
    area: str
    type: str
    vendor: str
    model: str
    serial_number: Optional[str] = None
    license_details: Optional[str] = None
    quantity: str = "1"
    sop_status: str = "Pending"
    email: Optional[str] = None
    phone: Optional[str] = None
    license_applicable: str = "No"
    account_type: str = "AUTO"
    security_level: str = "LOW"
    web_support: Optional[str] = None
    username: Optional[str] = None
    credentials: Optional[str] = None
    otp_required: Optional[str] = None
    contact_person_otp: Optional[str] = None
    validity: Optional[str] = None
    contact_info: Optional[str] = None
    contact_number: Optional[str] = None


class EquipmentUpdate(BaseModel):
    name: Optional[str] = None
    area: Optional[str] = None
    type: Optional[str] = None
    vendor: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None
    license_details: Optional[str] = None
    quantity: Optional[str] = None
    sop_status: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    license_applicable: Optional[str] = None
    account_type: Optional[str] = None
    security_level: Optional[str] = None
    web_support: Optional[str] = None
    username: Optional[str] = None
    credentials: Optional[str] = None
    otp_required: Optional[str] = None
    contact_person_otp: Optional[str] = None
    validity: Optional[str] = None
    contact_info: Optional[str] = None
    contact_number: Optional[str] = None


class EquipmentResponse(BaseModel):
    id: int
    name: str
    area: str
    type: str
    vendor: str
    model: str
    serial_number: Optional[str] = None
    license_details: Optional[str] = None
    quantity: str
    sop_status: str
    email: Optional[str] = None
    phone: Optional[str] = None
    license_applicable: str
    account_type: str = "AUTO"
    security_level: str = "LOW"
    web_support: Optional[str] = None
    username: Optional[str] = None
    credentials: Optional[str] = None
    otp_required: Optional[str] = None
    contact_person_otp: Optional[str] = None
    validity: Optional[str] = None
    contact_info: Optional[str] = None
    contact_number: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    manual: Optional[ManualContentResponse] = None
    attachments: List[AttachmentResponse] = []

    class Config:
        from_attributes = True


class EquipmentListResponse(BaseModel):
    id: int
    name: str
    area: str
    type: str
    vendor: str
    model: str
    quantity: str
    sop_status: str
    email: Optional[str] = None
    phone: Optional[str] = None
    account_type: str = "AUTO"
    security_level: str = "LOW"
    doc_count: int = 0
    video_count: int = 0
    last_updated: Optional[datetime] = None

    class Config:
        from_attributes = True

File: client/src/api/equipment.ts
import { api } from './client';
import type { Equipment, EquipmentListItem, ManualContent, Attachment, DocumentCategory } from '../types';

export interface EquipmentCreateInput {
  name: string;
  area: string;
  type: string;
  vendor: string;
  model: string;
  serial_number?: string;
  license_details?: string;
  quantity?: string;
  sop_status?: string;
  email?: string;
  phone?: string;
  license_applicable?: string;
  web_support?: string;
  username?: string;
  credentials?: string;
  otp_required?: string;
  contact_person_otp?: string;
  validity?: string;
  contact_info?: string;
  contact_number?: string;
}

export interface ManualContentInput {
  summary: string;
  monitoring: string[];
  maintenance: string[];
  troubleshooting: string[];
  links?: { title: string; uri: string }[];
  illustration_prompt?: string;
  image_url?: string;
}

export const equipmentApi = {
  getAll: () => api.get<EquipmentListItem[]>('/equipment/'),

  getById: (id: number) => api.get<Equipment>(`/equipment/${id}`),

  create: (data: EquipmentCreateInput) =>
    api.post<Equipment>('/equipment', data),

  update: (id: number, data: Partial<EquipmentCreateInput>) =>
    api.put<Equipment>(`/equipment/${id}`, data),

  delete: (id: number) => api.delete(`/equipment/${id}`),

  saveManual: (equipmentId: number, manual: ManualContentInput) =>
    api.put<Equipment>(`/equipment/${equipmentId}/manual`, manual),

  generateManual: (equipmentId: number) =>
    api.post<ManualContent>(`/manuals/generate/${equipmentId}`),
};

export interface UploadAttachmentInput {
  equipmentId: number;
  file: File;
  fileType: string;
  documentCategory: DocumentCategory;
}

export interface AddUrlAttachmentInput {
  equipmentId: number;
  name: string;
  url: string;
  fileType: 'youtube' | 'web';
  documentCategory: DocumentCategory;
  metadata?: Record<string, string>;
}

export const attachmentsApi = {
  upload: async (input: UploadAttachmentInput): Promise<Attachment> => {
    const formData = new FormData();
    formData.append('file', input.file);
    formData.append('file_type', input.fileType);
    formData.append('document_category', input.documentCategory);

    const response = await fetch(`/api/attachments/${input.equipmentId}`, {
      method: 'POST',
      credentials: 'include',
      body: formData,
    });

    if (!response.ok) {
      throw new Error('Upload failed');
    }

    return response.json();
  },

  addUrl: async (input: AddUrlAttachmentInput): Promise<Attachment> => {
    const formData = new FormData();
    formData.append('name', input.name);
    formData.append('url', input.url);
    formData.append('file_type', input.fileType);
    formData.append('document_category', input.documentCategory);
    if (input.metadata) {
      formData.append('metadata', JSON.stringify(input.metadata));
    }

    const response = await fetch(`/api/attachments/${input.equipmentId}/url`, {
      method: 'POST',
      credentials: 'include',
      body: formData,
    });

    if (!response.ok) {
      throw new Error('Add URL failed');
    }

    return response.json();
  },

  togglePublish: (attachmentId: number) =>
    api.patch<{ id: number; is_published: boolean }>(`/attachments/${attachmentId}/publish`),

  delete: (attachmentId: number) =>
    api.delete(`/attachments/${attachmentId}`),
};
